test/fuzz/lua/mvcc_test.lua

```diff
@@ -0,0 +1,756 @@
+                       "]]\n" ..
+                       "\n"):format(msg))
+    serial_file:write(generate_init_code(false) ..
+                             generate_schema_code())
+    for _, stmt in ipairs(serial) do
+        serial_file:write(generate_txn_serial_stmt(stmt))
+    end
+    serial_file:write("\n" ..
+                      "os.exit()\n")
+end
+
+local function tx_call(tx, operation)
+    local ok, res = pcall(tx._strm.eval, tx._strm, 'return ' .. operation.str)
+    table.insert(stmts, {
+        tid  = tx.id,
+        type = operation.type,
+        str  = operation.str,
+        ok   = ok,
+        res  = res,
+    })
+    if ok and operation.type == DML then tx.ro = false end
+    if not ok then
+        if res.message == 'Transaction has been aborted by conflict' then
+            tx:rollback()
+            return ok, res
+        end
+        if operation.type == DML then tx.bad_dml = true end
+    end
+    if tx._strm._conn.state == 'error' then
+        dump_repro()
+        t.fail(('connection in error state: %s'):format(tx._strm._conn.error))
+    end
+    return ok, res
+end
+
+local function tx_begin(tx)
+    t.fail_if(tx.running or tx.committed or tx.aborted,
+              'internal test error: cannot start transaction')
+    tx{
+        tid  = tx.id,
+        type = TXL,
+        str  = 'box.begin()',
+    }
+    tx.running = true
+end
+
+local function tx_rollback(tx)
+    t.fail_if(not tx.running or tx.committed or tx.aborted,
+              'internal test error: cannot rollback transaction')
+    local ok, res = tx{
+        tid  = tx.id,
+        type = TXL,
+        str  = 'box.rollback()',
+    }
+    if not ok then t.fail('`stream:rollback` failed: ' .. res[1].error) end
+    tx.running = false
+    tx.aborted = true
+end
+
+local function tx_commit(tx)
+    t.fail_if(not tx.running or tx.committed or tx.aborted,
+              'internal test error: cannot commit transaction')
+    local ok, err = tx{
+                tid  = tx.id,
+                type = TXL,
+                str  = 'box.commit()',
+    }
+    if not ok then
+        t.fail_if(err.message ~= 'Transaction has been aborted by conflict',
+                  '`stream:commit` failed unexpectedly: ' .. err)
+        tx.running = false
+        tx.aborted = true
+        tx.str = 'box.rollback()'
+        return false
+    end
+    tx.running = false
+    tx.committed = true
+    return true
+end
+
+local function tx_new(conn, id)
+    local mt = {
+        __index = {
+            begin    = tx_begin,
+            rollback = tx_rollback,
+            commit   = tx_commit,
+        },
+        __call = tx_call,
+    }
+    return setmetatable({
+        id        = id,
+        _strm     = conn:new_stream(),
+        running   = false,
+        committed = false,
+        aborted   = false,
+        ro        = true,
+        bad_dml   = false,
+    }, mt)
+end
+
+local function gen_key_or_null()
+    if math.random() <= p_null_key then
```

Comment:
I would define a function `rand_number()` with `math.random` inside. It allows us to easily redefne such functions for using with real fuzzing engines that generate values from FDP.